Update MVP+Makefile.md (#14)
IshaanAdarsh/Postgres-extension-tutorial: Quick-Start [DA]/MVP/MVP+Makefile.md

# Minimum Viable Product:
A basic tutorial on creating a PostgreSQL extension using the bare minimum files and a Makefile:

- A **control file**, which tells PostgreSQL some basic information about the extension, such as its name, version, and schema.
- A **SQL script file**, which contains the SQL commands to create the extension's objects.

> - A **Makefile** is an optional but highly recommended file that helps automate the build process of the extension. It defines the compilation and installation steps, making it easier to build and install the extension on different systems.

Reviewer: The Makefile is technically optional, but I don't think I'd mention that.

Author: Sure i'll change it.

# [CODE](https://github.com/IshaanAdarsh/Postgres-extension-tutorial/tree/main/Code/my_extension)

## A Postgres extension using only a control file, an SQL file and a Makefile:

### Step 1: Create the Extension Directory
- Create a directory for your extension. Let's call it "my_extension".

```bash
$ mkdir my_extension
$ cd my_extension
```

### Step 2: Create the Extension Control File
- Inside the "my_extension" directory, create a file named "my_extension.control". 
  - This file defines the metadata and dependencies of your extension.

```bash
$ touch my_extension.control
```

- Open the "my_extension.control" file and add the following content:

> ```control

Reviewer: I wonder if `ini` syntax would work well here.

Author: ERROR: does not adhere to the INI syntax
I tried using `ini` syntax in control file, got this error in my iterm shell

# my_extension.control
comment = 'Minimal Viable Product'
default_version = '1.0'
relocatable = true

# Installation script for version 1.0
# 1.0
# my_extension--1.0.sql
```

> 

Reviewer: I suggest adding some notes about what these do and a link to the docs.

Author: I will make a /doc\README.md for all the details for the extension. 
I want to package and release it on PGXN, after that all the changes i perform i will update in the extension package

### Step 3: Create the Extension SQL File
- Create a file named "my_extension--1.0.sql".
  - This file contains the SQL statements required to install or uninstall your extension.

```bash

> $ touch my_extension--1.0.sql

Reviewer: Mention that the double-dash is meaningful so that Postgres know which part of the file name is the version number.

Author: Yes, i will make those changes

```

- Open the "my_extension--1.0.sql" file and add the following content:

```sql

> -- extension--1.0.sql

Reviewer: Old line?

Author: yup

-- my_extension--1.0.sql
-- Create necessary objects for version 1.0

> CREATE TABLE my_table (

Reviewer: I suggest starting with a function rather than a table. Something as simple as this will do:

``` sql
CREATE FUNCTION add_two(
    n int
) RETURNS INT LANGUAGE SQL AS 'SELECT n + 2';
```

Author: I had a funtion in the previous code, but due to some issue, it was not working in the regression testing.
So i switched to the `CREATE TABLE`

I will figure out where i went wrong and will add a function to the SQL file as well

  id SERIAL PRIMARY KEY,
  name VARCHAR(100) NOT NULL
);
```

### Step 4: Create the Makefile
- Create a Makefile in the "extension" directory. 

>   - The Makefile is used to define the build process. It specifies the module name, object files, extension name, and SQL files. The PG_CONFIG variable retrieves the path to the pg_config utility, which is used to get the necessary configuration for building the extension.

Reviewer: Omit mention of the object file here, since this extension doesn't have one.

```bash
$ touch Makefile
```
- Open the Makefile and add the following content:

> <!--Problem: https://dba.stackexchange.com/questions/328624/makefile-issue-in-creating-postgres-extension

Reviewer: Since you've solved this problem, do you need to keep this comment?

Author: These comments are just for my reference, just to make sure that if i run into these issues again i know what the problem and the solution is

Had an issue with the make install command:
```makefile
DATA = my_extension--1.0.sql

PG_CONFIG  ?= pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
```
Solution: Makefile is lacking the entry `EXTENSION = my_extension`
Reason: That's why `make install` doesn't copy the `*.control` file, and that's why `CREATE EXTENSION` fails.

NOTE: make: Nothing to be done for 'all' 
Makefile doesn't have any MODULE or other binary to build. You can go directly to the make install step
-->

```makefile
EXTENSION = my_extension
DATA = my_extension--1.0.sql

PG_CONFIG  ?= pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)

> include $(PGXS)

Reviewer: Say something here about how PGXS is a library of Makefile commands to simplify building and installing extensions just by setting variables it reads, like `EXTENSION` and `DATA`.

Author: I will add a comment expalining each line of the code.
I will add another section on PGXS and PGXN after regression testing, to cover all bases

Reviewer: Okay, although PGXN doesn't really enter into this example at all.

Author: I will just explain PGXN in 1-2 sentences, its use basically.
Then i want to proceed with the Extension management and upgradability.
That should cover all bases.

```

### Step 5: Build and Install the Extension
- To build and install the extension, run the following commands:

```bash
$ make
$ make install
```

> - The "make" command compiles the extension code using the Makefile

Reviewer: This is true, although here there is nothing to compile.

Author: yup I mentioned this in the comments, I will add a line that specifies in this particular extension code there is No Modules to install, so make doesn't serve a purpose.

- The "make install" command copies the necessary files to the appropriate PostgreSQL directories.

### Step 6: Create the Extension in PostgreSQL
- Connect to your PostgreSQL database and execute the following SQL statement to create the extension:

```sql
CREATE EXTENSION IF NOT EXITS "my_extension";

> ```

Reviewer: I suggest you show the output of all the commands you run.

- The CREATE EXTENSION statement is executed in PostgreSQL to create the extension.

## Implementation of the Extension:
Once you have successfully implemented the `my_extension` in your PostgreSQL database, you can use it as follows:

### 1. Query the `my_table` table:
   - You can query the `my_table` table like any other table in your database. For example, you can retrieve all rows from the table using the following SQL query:
   ```sql
   SELECT * FROM my_table;
   ```
#### Output:

> ```sql

Reviewer: This output is not SQL; use plain text formatting.

id | name
----+------
(0 rows)
```

> Remember to replace `my_table` with the actual table name you have chosen, and adjust the function call as needed based on your specific implementation.

Reviewer: I don't understand this sentence, since you didn't have the reader change the table name, nor is there a function call. Leftover from a previous doc?

Author: the first part of the point is faulty i will look into it

yup, the function call is something I have to fix.
When I fix the function, the 2nd part of the line will make sense.

## Regression Testing:

### Step 1: Create the Extension SubDirectory for Regression SQL File:
- Create a subdirectory for your regression file. Name it `sql`

```bash
$ mkdir sql
$ cd sql
```
-   Create a new file called `my_extension_regression.sql`.

```bash
$ touch my_extension_regression.sql
```

- Open the `my_extension_regression.sql` file and add the following content keeping in mind these instructions:

  -   Write SQL statements that test the functionality of your `my_extension`.

  -   Include a combination of `CREATE`, `INSERT`, `UPDATE`, `DELETE`, and `SELECT` statements to cover various scenarios.

  -   Make sure to test both positive and negative scenarios to validate the behaviour of your extension.

```sql
-- regression test script for my_extension
-- Create the my_table table

> CREATE TABLE my_table (

Reviewer: Why would you be creating this table when the extension already creates it? The first thing the script should do is `CREATE EXTENSION my_extension`.

Author: Fixed it

-- Verify the data in the my_table table
SELECT * FROM my_table;
```

### Step 2: Update the Makefile:

- In the given code, the changes made to the `Makefile` are related to the addition of a new variable called `REGRESS` and the inclusion of a new SQL file `my_extension--regress`.

```makefile
EXTENSION = my_extension
DATA = my_extension--1.0.sql
REGRESS = my_extension--regress

PG_CONFIG  ?= pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
```

- The `REGRESS` variable is used to specify the name of the regression test script file for the extension. In this case, the name is `my_extension--regress`. This file will contain the SQL commands for the regression tests to be executed.

- These changes in the `Makefile` enable the `my_extension--regress.sql` file to be recognized during the installation process, allowing the regression tests to be executed using `make installcheck` 

### Step 3: Create the Extension SubDirectory for Expected Result:
- Create a subdirectory for your regression file. Name it `expected`

```bash
$ mkdir expected
$ cd expected
```
-   Create a new file called `my_extension--regress.out`.

```bash
$ touch my_extension--regress.out
```

> - Add your assumption of what will be the result of the regression test:

Reviewer: Mention that it should be *exactly* the expected text from the script.

```out
-- regression test script for my_extension
-- Create the my_table table
CREATE TABLE my_table (
  id SERIAL PRIMARY KEY,
  name VARCHAR(100) NOT NULL
);
-- Delete existing rows from my_table
DELETE FROM my_table;
-- Verify the data in the my_table table
SELECT * FROM my_table;
 id | name 
----+------
(0 rows)

```

> - If the expected file matched the results, the teat is passed.

Reviewer: ```suggestion
- If the expected file matched the results, the teat passed.
```

### Step 4: Run `make installcheck`:

- Navigate to the directory containing your extension's code and Makefile.
- Run the following command to execute the `make installcheck` target:
   ```
   make installcheck
   ```

   This command will initiate the installation and regression testing process for your extension.

#### Working of the command `make installcheck`:
- The `make installcheck` command will internally execute the `pg_regress` utility with the appropriate parameters. It will connect to a PostgreSQL server and run the regression tests specified in the `my_extension--regress.sql` file.

- Review the output of the `make installcheck` command to see the results of the regression tests. The command will display the progress and status of each test case.

>   - If all the tests pass successfully, you will see a summary indicating that the regression tests have passed.

Reviewer: Include output to show what that looks like.

Reviewer: Oh, you have it below, never mind!

>   - If any tests fail, the output will provide information about the specific test case that failed and any relevant error messages.

Reviewer: It will also create a diff file showing the difference between actual and expected output. One should include this file in one's `.gitignore` file.

```

For more detailed information on Control Files, you can refer to the [PostgreSQL documentation on Control Files](https://www.postgresql.org/docs/current/extend-extensions.html#id-1.8.3.20.11).

### Step 3: Create the Extension SQL File
- Create a file named "my_extension--1.0.0.sql".
  - This file contains the SQL statements required to install or uninstall your extension.
  - The format for a PostgreSQL extension file name is `extension_name--version.sql`, where `extension_name` is the name of the extension, and `version` is the version number. The double-dash serves as a delimiter to indicate the separation between the extension name and the version number.

```bash
$ touch my_extension--1.0.0.sql
```
- Open the "my_extension--1.0.0.sql" file and add the following content:

```sql
-- Create necessary objects for version 1.0.0
CREATE TABLE my_table (
  id SERIAL PRIMARY KEY,
  name VARCHAR(100) NOT NULL
);

-- Add 2 numbers using add function
CREATE FUNCTION add(a integer, b integer) RETURNS integer
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT
    RETURN a + b;
```

### Step 4: Create the Makefile
- Create a Makefile in the "extension" directory. 

>   - The Makefile is used to define the build process. It specifies the module name, extension name, and SQL files. The PG_CONFIG variable retrieves the path to the pg_config utility, which is used to get the necessary configuration for building the extension.

Author: Module specifies the modules to bt installed using the `make` command. As this is a SQL Extnesion we don't really need a module but while using C we need to specify module to create the `.so` and `.o` files. I have added this as a further explanation of the make file error of sort below.

>   - The Makefile is used to define the build process. It specifies the module name, extension name, and SQL files. The PG_CONFIG variable retrieves the path to the pg_config utility, which is used to get the necessary configuration for building the extension.

Reviewer: Yeah, since there is no C here perhaps don't venture module. But also, it's `module_pathname`.

>   - The Makefile is used to define the build process. It specifies the module name, extension name, and SQL files. The PG_CONFIG variable retrieves the path to the pg_config utility, which is used to get the necessary configuration for building the extension.

Author: yup i'll remove it.

```bash
$ touch Makefile
```
- Open the Makefile and add the following content:

```makefile
EXTENSION = my_extension                 # Name of the extension
DATA = my_extension--1.0.0.sql           # SQL file containing extension objects and functions

PG_CONFIG  ?= pg_config                  # Path to the pg_config executable
PGXS := $(shell $(PG_CONFIG) --pgxs)     # Get the PostgreSQL Extension Makefile from pg_config
include $(PGXS)                          # Include the PostgreSQL Extension Makefile
```

> 

Reviewer: Yes, it's enough to explain what these 3-4 lines do at a high level, I think.

3. Inside the Makefile, include PGXS by adding a few lines of code.
4. Define any extension-specific instructions in the Makefile.
5. Use the `make` command to build and install your extension.
6. Enable the extension in PostgreSQL using the `CREATE EXTENSION` SQL command.

PGXS takes care of locating your PostgreSQL installation, determining compilation options, and ensuring the extension files are in the right place. It simplifies the process, making it easier for you to build and install your extension in different PostgreSQL environments.
-->

For more information on the Makefiles, you can refer to the [PostgreSQL documentation on makefile](https://www.postgresql.org/docs/15/extend-pgxs.html).

#### Tip: 
```makefile
EXTENSION = my_extension
DATA = my_extension--1.0.0.sql my_extension--1.1.0.sql
```

- By including multiple SQL script files in the `DATA` variable of the `Makefile`, the user can ensure that different versions of the extension can have different sets of objects or functionality. 
- Each SQL script file represents a specific version of the extension and can contain the necessary SQL commands for the desired changes.

> - This allows the user to easily manage and control the installation of specific features or modifications in different versions of your extension.

Reviewer: It took me a few minutes to parse this section. I think the important point is that different versions of an extension can be supported by having files for each version, e.g., `my_extension-1.0.0.sql`, `my_extension-1.0.1.sql`, `my_extension-1.2.0.sql`, etc. You tell Postgres about them by including them all in the `DATA` variable, either as a list or using wildcards, e.g., `DATA = $(wildcard my-extension--*.sql)`.

Author: I was suggested this by a fellow gsoc contributor when i asked him for some feedback. He told me that he had a problem with understanding this part. Most people know about the wildcard but including the 2-3 SQL scripts to load all of them isn't mentioned probably.

I can add a section about wildcard to give a better approach for having all the versions.

Reviewer: I think it's fine to explicitly list the files as an example; my point is that I had a hard time understanding your explanation.

Author: ohk, i will change it up and submit it for review again.

### Step 5: Build and Install the Extension

> - To build and install the extension, run the following commands:

Reviewer: I think it's worth covering the need for`make` and PGXS, often installed via `-lib` packages in Yum, apt, and the like.

Author: I can cover the `make` explaination in the use of external files( C extensions ). Its wont be helpful for the user to see the whole explanation all here with out an example to back it.

I can add a whole section on using PGXN and add a link to this original QS. I was thinking that this thould be the main QS and all the other QS be linked to this central basic QS. Currently we have:
- QS on explaining Extensibility
- Procedural Language (plpython) [Need to add plpgsql]

- I will add a separate PGXN QS, which cover `pgxnclient`. It can link to the PGXN docs too.
- External language is also left we will need to cover that as well (Python and C)

Reviewer: I'm not referring to PGXN, but that the use of a `Makefile` requires `make` and the use of `PGXS` (not PGXN*N*) requires the Postgres development libraries.

Author: **Need for `make`:**
   - `make` is a build automation tool used to compile and build software projects, including PostgreSQL extensions.
   - It reads instructions from a file called "Makefile" to determine the build process, dependencies, and rules for compilation.
   - Without `make`, developers would need to manually manage complex compilation steps, leading to errors and inefficiencies. 

**How `make` Command Functions:**
   - The `make` command is executed in the extension's directory to initiate the build process defined in the Makefile.
   - It automatically compiles source files, links object files, and generates the shared object (`.so`) file for the extension.
   - By default, `make` uses the `all` target in the Makefile, but developers can specify other targets for specific tasks, such as cleaning or installing the extension.

**The Role of PGXS (PostgreSQL Extension Build Infrastructure):**
   - PGXS is a standardized build infrastructure provided by PostgreSQL for developing extensions independently of the core codebase.
   - It includes macros, rules, and utilities that streamline the extension development process and ensure compatibility with the installed PostgreSQL version.
   - PGXS simplifies the integration of extensions with the PostgreSQL installation, making it easier to distribute and manage extensions.

**Installation and Use of PGXS:**
   - The PGXS infrastructure is installed with the PostgreSQL development libraries and headers.
   - On Linux distributions, developers can install the necessary packages, such as `postgresql-server-dev-X`, via package managers like Yum or Apt.
   - PGXS provides the necessary environment and tools for extension development, ensuring compatibility with the installed PostgreSQL version. 

**Using a Makefile for Extension Building:**
   - When building a PostgreSQL extension, developers create a Makefile with instructions for compiling and linking extension code.
   - The Makefile uses PGXS macros to include PostgreSQL-specific build settings, libraries, and header files necessary for extension development.
   - The Makefile specifies the extension's shared object (`.so`) file, name, and version, ensuring proper integration with the PostgreSQL environment.

these are some of the points i could get from the previous reference documentation i made. @theory Could you guide me with points should i include?

Reviewer: If you still have reference documentation, it'd be good to simply link to these bits, but honestly I think your readers will be best served by a list of instructions for how to install all the dependencies on RedHat-based Linux, Debian-based Linux, macOS, and Windows.

### Step 6: Create the Extension in PostgreSQL:
- Connect to your PostgreSQL database and execute the following SQL statement to create the extension:

```sql
CREATE EXTENSION IF NOT EXISTS my_extension;
-- Output: CREATE EXTENSION
```
- The CREATE EXTENSION statement is executed in PostgreSQL to create the extension.

For more information on the `CREATE EXTENSION` command, you can refer to the [PostgreSQL documentation on `CREATE EXTENSION`](https://www.postgresql.org/docs/current/sql-createextension.html).

## Implementation of the Extension:
Once you have successfully implemented the `my_extension` in your PostgreSQL database, you can use it as follows:

### 1. Query the `my_table` table:
   - You can query the `my_table` table like any other table in your database. For example, you can retrieve all rows from the table using the following SQL query:
   ```sql
   SELECT * FROM my_table;
   ```
#### Output:
```
id | name
----+------
(0 rows)
```
### 2. Use the add function to perform addition:
  - The `add` function takes two integer arguments and returns their sum
  ```sql
  SELECT add(2, 3);
  ```
#### Output:
```
 add
-----
   5
(1 row)
```

## Regression Testing:

> ### Step 1: Create the Extension SubDirectory for Regression SQL File:

Reviewer: ```suggestion
### Step 1: Create the Extension subdirectory for test SQL files:
```

- Create a subdirectory for your regression file. Name it `sql`

```bash
$ mkdir sql
$ cd sql
```

> -   Create a new file called `my_extension--regress.sql`.

Reviewer: I recommend using a single dash, here; double dashes are used in extension file names because they have special meaning, separating names from versions.

Author: Sure i will make the changes.

Reviewer: I don't understand the comment. You don't need the word `regression` in the file name.

Author: Okay, i miss understood. I just was using a particular extension as an example. He used --regress.sql, so i followed suit.

```bash
$ touch my_extension--regress.sql
```

- Open the `my_extension--regress.sql` file and add the following content keeping in mind these instructions:

  -   Write SQL statements that test the functionality of your `my_extension`.

>   -   Include a combination of `CREATE`, `INSERT`, `UPDATE`, `DELETE`, and `SELECT` statements to cover various scenarios.

Reviewer: Perhaps then link to some examples?

It's important to consider the specific requirements, compatibility, and impact on dependencies when choosing the appropriate method for upgrading extensions in PostgreSQL.

## Updating the Extension using the `ALTER EXTENSION` command:

- The ALTER EXTENSION command in PostgreSQL enables users to efficiently modify extensions by upgrading to newer versions, adjusting configuration and behaviour, introducing additional functionality or objects, and modifying existing components within the extension. 

- This command streamlines extension management, eliminating the need for manual interventions or complex administrative procedures. It simplifies the task of applying updates, incorporating new features, and ensuring compatibility with evolving requirements.

- For more information on the Extension Updates, you can refer to the [PostgreSQL documentation on Extension Updates](https://www.postgresql.org/docs/current/extend-extensions.html#id-1.8.3.20.14)

### Directory Hierarchy:
```
my_extension/
├── README.md
├── my_extension--1.0.0.sql

> ├── my_extension--1.0.0--1.0.1.sql

Author: For users who are installing the extension from scratch, including all SQL script files, such as `my_extension--1.0.0.sql` and `my_extension--1.0.1.sql`, ensures that they can choose the specific version they desire. By having both SQL script files available, users can set up the extension with the necessary objects and functionality corresponding to the version they want to install. This flexibility allows users to easily tailor the extension to their specific requirements, enhancing their overall experience.

i could add this point just to clarify.

-- Create necessary objects for version 1.0.1
-- CREATE TABLE my_table (
--  id SERIAL PRIMARY KEY,
--  name VARCHAR(100) NOT NULL
-- );

-- Add 2 numbers using add function
-- CREATE FUNCTION add(a integer, b integer) RETURNS integer
--    LANGUAGE SQL
--    IMMUTABLE
--    RETURNS NULL ON NULL INPUT
--    RETURN a + b;

-- Add complex_add function in version 1.0.1
CREATE FUNCTION complex_add(integer[]) RETURNS integer
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT
    AS $$
        SELECT COALESCE(SUM(val), 0)
        FROM unnest($1) AS t(val)
    $$
;
```

### Step 2: Update the Control File:
- Updating the default_version for an extension is necessary to indicate the latest version available

```control
# my_extension.control
comment = 'Minimal Viable Product'
default_version = '1.0.1'                # Reflect the newer version of the extension
relocatable = true
```

### Step 3: Update the Makefile:
- Update the `DATA` variable to include the update script to ensure that it is executed during the upgrade process of the extension, enabling the changes to be applied consistently across all installations.

```makefile
EXTENSION = my_extension

> DATA = my_extension--1.0.0--1.0.1.sql   # Update the DATA variable with the Upgrade Script file

Author: Yup. I will add that.

> DATA = my_extension--1.0.0--1.0.1.sql   # Update the DATA variable with the Upgrade Script file

Author: ```suggestion
DATA = my_extension--1.0.0.sql my_extension--1.0.0--1.0.1.sql   # Update the DATA variable with the Upgrade Script file
```

REGRESS = my_extension--regress

PG_CONFIG  ?= pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
```

### Step 4: Build and Install the Extension
- To build and install the updated extension, run the following commands:

```bash
$ make
$ make install
```

### Step 5: Upgrade the Extension:
- Finally, use the `ALTER EXTENSION` command to update the extension to the latest version.
- Execute the following SQL statement:

```sql
ALTER EXTENSION my_extension UPDATE TO '1.0.1';
-- Output: ALTER EXTENSION
```
The new objects introduced in the `my_extension--1.0.0--1.0.1.sql` file are included when updating an existing extension. This approach allows you to add new functionality without affecting the existing objects and ensures proper versioning and upgrade procedures.

- For more information on the `ALTER EXTENSION` command, you can refer to the [PostgreSQL documentation on `ALTER EXTENSION`](https://www.postgresql.org/docs/current/sql-alterextension.html)

- To check for **unexpected update paths**, use this command:

```sql

> SELECT * FROM pg_extension_update_paths('*`extension_name`*');

Author: no its a typo error

> SELECT * FROM pg_extension_update_paths('*`extension_name`*');

Author: ```suggestion
SELECT * FROM pg_extension_update_paths('extension_name');
```

```bash
$ touch my_extension--1.0.1.sql
```

- Open the `my_extension--1.0.1.sql` file and add the following content:
  - The `complex_add` is a new function named complex_add that takes an array of integers as input. Inside the function, we use the unnest function to expand the array into individual elements and then calculate the sum using the SUM function. The result is returned as an integer.

```sql
-- Update path script for version 1.0.1
-- Create necessary objects for version 1.0.1
CREATE TABLE my_table (
  id SERIAL PRIMARY KEY,
  name VARCHAR(100) NOT NULL
);

-- Add 2 numbers using add function
CREATE FUNCTION add(a integer, b integer) RETURNS integer
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT
    RETURN a + b;

CREATE FUNCTION complex_add(integer[]) RETURNS integer
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT
    AS $$
        SELECT COALESCE(SUM(val), 0)
        FROM unnest($1) AS t(val)
    $$
;
```
#### 3) Update the Makefile:

- In the given code, the changes made to the `Makefile` are related to the updated value of the `DATA` variable.

```makefile
EXTENSION = my_extension

> DATA = my_extension--1.0.1.sql

Author: Completed the changes.

REGRESS = my_extension--regress

PG_CONFIG  ?= pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
```

### Step 4: Install the new version: 
- Run the `CREATE EXTENSION` command to install the updated version of the extension:
```sql
CREATE EXTENSION your_extension_name;
-- Output: CREATE EXTENSION
```

For more information on the `CREATE EXTENSION` command, you can refer to the [PostgreSQL documentation on `CREATE EXTENSION`](https://www.postgresql.org/docs/current/sql-createextension.html).

### Step 5: Verify the upgrade: 
- Confirm that the upgrade was successful by checking the new version of the extension:
```sql

> SELECT * FROM pg_extension WHERE extname = 'your_extension_name';

Author: ```suggestion
SELECT * FROM pg_extension WHERE extname = 'my_extension';
```

  - Introduction
  - Installation Instructions
  - Configuration Options
  - Usage Examples
  - Troubleshooting and FAQ
  - License and Copyright Information.

```md
## Introduction
my_extension is a minimal viable product that provides a set of functions and a table to store data in PostgreSQL.

## Installation Instructions
To install my_extension, follow these steps:

1. Ensure you have PostgreSQL installed on your system.

2. Download the my_extension files.

3. Navigate to the directory containing the my_extension files.

4. Build and install the extension using the following command:

  `make install`
  

5. Enable my_extension in your PostgreSQL database:

  `CREATE EXTENSION my_extension;`

## Configuration Options

my_extension does not require any specific configuration options.

## Usage Examples

### Creating and Using the Table

To create the necessary objects for version 1.0.0 and the my_table table:

  `CREATE TABLE my_table (`
  `  id SERIAL PRIMARY KEY,`
  `  name VARCHAR(100) NOT NULL`

>   `);`

Author: The README.md file is inside the code block. If i am using these i am separating the code books. If there is a way to nest code block i could apply that but i looked around and there is no ways to do that.  

Its just to emphasise that it is a Code block.

  - Architecture and Design
  - Functionality Explanation
  - Code Structure
  - Performance Considerations
  - Testing and Regression
  - Contribution Guidelines

```md
## Architecture and Design

my_extension provides a simple set of functions and a table in PostgreSQL to demonstrate a minimal viable product. It consists of a single table, "my_table," with two columns: "id" (a serial primary key) and "name" (a varchar).

## Functionality Explanation

### Table: my_table
The "my_table" table is used to store data with an auto-incrementing "id" and a "name" field to hold a string value.

### Function: add
The "add" function is a simple SQL function that takes two integer inputs and returns their sum.

### Function: complex_add
The "complex_add" function is an additional function introduced in version 1.0.1. It takes an integer array as input and returns the sum of its elements.

> ## Code Structure:

Author: I had a doubt regarding this. There are a lot of parts which are repeated in the README.md and docfile. Shouldn't there be a distinction between these two?

Moreover, could you just direct me on which parts should i include in the readme and what to include in the doc file. I will make the changes accordingly

## Version 1.0.1

### Added
- Introduced the `complex_add` function to perform sum calculations on an integer array input.
- Extended the my_extension functionality to include more advanced arithmetic operations.
- Updated the SQL file to include the definition of the `complex_add` function.

```

## Extension Distribution:
Once you have developed and tested your Postgres extension, it's time to distribute it to other users and developers. Distributing your extension ensures that others can benefit from your work and integrate it into their PostgreSQL databases. To achieve this, you can use various platforms and distribution methods to make your extension accessible to a diverse audience.

### 1. PGXN (PostgreSQL Extension Network)

PGXN is a centralized repository for PostgreSQL extensions, providing a convenient platform for extension distribution. By submitting your extension to PGXN, you can make it readily available to users and developers in the PostgreSQL community.

#### Steps to Publish Your Extension on PGXN:

1. Ensure your extension adheres to PGXN's packaging requirements and standard extension structure.

> 2. Create a distribution package (tarball or zip) of your extension with all the necessary files, including the `.control`, `META.json`, SQL scripts, and documentation. For more detailed information on how to create a distribution package visit the [PGXN How To](https://manager.pgxn.org/howto)

Author: I will link to the `PGXN and pgxnclient` #31  explanation for this. It doesn't make sense to repeat the steps separately.

3. Register an account on [PGXN](https://pgxn.org).

> 4. Use the PGXN client tool to package and publish your extension:

Reviewer: There is a lot more to it than this. Releasing is not included in the [pgxnclient](https://pypi.org/project/pgxnclient/) ([repo](https://github.com/pgxn/pgxnclient)). For that you also need to install [pgxn-utils](guedes/pgxn-utils), which adds the command to upload to PGXN.

Author: `pgxn_utils` has depreciated as far as i know. It has not been updated in the last two years. Should i include it in this QS?

Reviewer: I haven't heard that it was deprecated; @guedes, do you continue to maintain `pgxn_utils`?

Reviewer 2: Hi all! 

I'd like to know whether `pgxn_utils` whorths yet to community. If you think it brings value to community yet, I can put some effort to update dependencies and review some issues/PRs. Furthermore, it could be nice to rewrite it with fewer dependencies and with more improvements in boilerplate/scaffold generation. 

I'm open to your comments :)

Reviewer: Well I certainly use it :-) Though in truth I use it less than I used to now that some of my extensions use workflow actions to release, [like this](https://github.com/theory/pg-semver/blob/main/.github/workflows/release.yml). But I've always been rather fond of `pgxn_utils` :-)

```bash

> pgxn load path/to/your_extension-1.0.0.zip

Reviewer: I believe is the command to `CREATE EXTENSION`. You first need to install it with `pgxn install my_extension`.

Author: I will remove all this part explaining in short on how to use PGXN and add a link to #31. We can make the changes there.

```

Replace `path/to/your_extension-1.0.0.zip` with the actual path to your extension's distribution package.

> 5. Your extension will now be available for others to discover and install using the PGXN client tool.

Reviewer: Yeah, not till it has been released on PGXN. I presume that's covered in another chapter, yes?

Author: yup

### 2. Package Managers

Utilizing package managers allows users to install your extension easily with a simple command. You can package your extension in formats compatible with various package managers like `apt`, `yum`, `brew`, or `pip`.

> #### Steps to Distribute Your Extension via Package Managers:

Reviewer: I think there  is a lot more to it than this brief section covers, including specifics about various package managers and how to get packages added to their repos. May not be worth including here for now.

Author: Just adding short explanations might not be helpful. I could just mention these ways and end there. 

Then anyone can look around and discover them according to their own volition.

Reviewer: Right, but if this is a getting started guide, the more it can help someone get from zero to building and testing without having to search the web for how to do something, the better!

1. Create a distribution package in the format required by the respective package manager. For example, `.deb` for `apt` on Debian-based systems, `.rpm` for `yum` on Red Hat-based systems, or a formula for `brew` on macOS.

2. Host the distribution package on a repository or server accessible to users. You can use platforms like GitHub Releases, PackageCloud, or a custom repository.

3. Users can then install your extension using the package manager's install command:

```bash
# For apt (Debian/Ubuntu):
sudo apt-get install your-extension

# For yum (CentOS/RHEL):
sudo yum install your-extension

# For brew (macOS):
brew install your-extension
```

### 3. PostgreSQL Extensions on Platforms

Various platforms provide extension management support for PostgreSQL, making it easier for users to find, install, and update extensions.
#### Distributing Your Extension via PostgreSQL Extensions on Platforms:

1. Submit your extension to PostgreSQL extensions platforms like Postgres.app for macOS or BigSQL for Windows and Linux.

2. Users can then install your extension directly from the platform's interface or command-line tools.

> ### 4. Custom Installation Script

Reviewer: This section could use a lot more information, as well, and I'm not sure there's much value in it here. What do you think?

Author: I think we should remove the steps and mention the ways of distributing only.
Explaining PGXN will be more than enough

Reviewer: It's less about PGXN than using `make`.

   Notes on upgrading to a new version and any migration steps required.
```

> 4.  **Build a Distribution Package:** Package your extension and the `META.json` file into a distribution package. This package should be in a compressed format (zip format) and include all the necessary files for installation and usage of your extension.

Reviewer: I honesty think the PGXN packaging should be a separate lesson.

Author: sure, i will make it a separate lesson

5.  **Publish Your Extension:** Use the PGXN command-line tool or web interface to publish your extension. Provide the distribution package created in the previous step and upload it to PGXN. The tool will validate the package and extract the metadata.

### Updating an Extension on PGXN:
#### New Directory Hierarchy:
```
my_extension/
├── changes
├── META.json
├── README.md
├── my_extension--1.0.0.sql
├── my_extension--1.0.1.sql
├── expected/
│   └── my_extension--regress.out
├── sql/
│   └── my_extension--regress.sql
├── my_extension.control
├── results/
│   └── my_extension--regress.out
└── Makefile
```

#### Step 1: Update the Extension Control File:
- Open the "my_extension.control" file and update the following content:

```control
# my_extension.control
comment = 'Minimal Viable Product'

> default_version = '1.0.1'

Reviewer: This isn't necessary to release on PGXN.

Author: I though you had to change all the files which point to the older version.
Does'nt that include the control file too, which points to the older version of the extension?

Reviewer: You only have to increment the distribution version if you've previously made a release on PGXN.

Author: I will mention that in the new PGXN quick start.

relocatable = true
```

#### Step 2: Create the new Extension SQL File:

- Create a file named `my_extension--1.0.1.sql`.

  - This file contains the new SQL statements for the extension.
  - The format for a PostgreSQL extension file name is `extension_name--version.sql`, where `extension_name` is the name of the extension, and `version` is the version number. The double-dash serves as a delimiter to indicate the separation between the extension name and the version number.

```bash
$ touch my_extension--1.0.1.sql
```

- Open the `my_extension--1.0.1.sql` file and add the following content:

>   - The `complex_add` is a new function named complex_add that takes an array of integers as input. Inside the function, we use the unnest function to expand the array into individual elements and then calculate the sum using the SUM function. The result is returned as an integer.

Reviewer: Expanding the extension with new functionality is a great lesson, but ought to be independent of distribution on PGXN, I think.

Author: What i'll do is i will keep the new function but remove all the information about the PGXN. I will cover that in a separate lesson.(I can give some information about PGXS too as it isn't too long)

```sql
-- Update path script for version 1.0.1
-- Create necessary objects for version 1.0.1
CREATE TABLE my_table (
  id SERIAL PRIMARY KEY,
  name VARCHAR(100) NOT NULL
);

-- Add 2 numbers using add function
CREATE FUNCTION add(a integer, b integer) RETURNS integer
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT
    RETURN a + b;

CREATE FUNCTION complex_add(integer[]) RETURNS integer
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT
    AS $$
        SELECT COALESCE(SUM(val), 0)
        FROM unnest($1) AS t(val)
    $$
;
```
### Step 3: Update the Makefile:

- In the given code, the changes made to the `Makefile` are related to the updated value of the `DATA` variable.

```makefile
EXTENSION = my_extension
DATA = my_extension--1.0.1.sql
REGRESS = my_extension--regress

PG_CONFIG  ?= pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
```

## Implementation of the Extension:

> Once you have successfully implemented the `my_extension` in your PostgreSQL database, you can use the new funtion as follows:

Reviewer: This is true if you install it and run `CREATE EXTENSION` where it did not exist before, but if it already exists the new function won't appear. That will require creating another file, `my_extension--1.0.0--1.0.1.sql` that contains just the new objects, including it in the `DATA` variable in the Makefile, and then, after installing, running `ALTER EXTENSION UPDATE`.

Author: Dropped the PGXN and added 2 ways of updating the extension:
- `ALTER EXTENSION`
- `DROP EXTENSION` and `CREATE EXTENSION`

I thinks the basic SQL quick start is pretty much done.

-- Create the extension
CREATE EXTENSION my_extension;
-- Verify the data in the my_table table
SELECT * FROM my_table;
 id | name 
----+------
(0 rows)

-- Test the add function
SELECT add(1, 2) AS result;
 result 
--------
      3
(1 row)

-- Test the complex_add function
SELECT complex_add(ARRAY[1, 2, 3, 4, 5]) AS result;
 result 
--------
     15
(1 row)
```

### Step 6: Create a change file:
- A changelog is a file that provides a comprehensive and organized list of changes, enhancements, and bug fixes for each release of your PostgreSQL extension. It serves as a communication tool between you and the users, helping them understand the modifications made and the impact they may have on their existing installations. 

- Contents for the my_extension changelog:
```change
# Changelog for my_extension

## Version 1.0.1
- New Function: complex_add
  - Description: Added a new function called complex_add.
  - Usage: complex_add takes an array of integers as input and returns their sum.
  - Example: SELECT complex_add(ARRAY[1, 2, 3, 4, 5]);

## Version 1.0.0
- Initial release of my_extension.
- Implemented basic functionality for data management and operations.
```

> ### Step 7: Update the `META.json`:

Reviewer: Did this file exist before?

Author: yes

- Update the version number in the META.json file to "1.0.1" and ensure that all other information, such as the abstract, description, and maintainer, is accurate and up to date.

```json 
{
  "name": "my_extension",
  "abstract": "A Basic PostgreSQL extension",
  "description": "my_extension is a PostgreSQL extension that provides additional functionality for data management and operations.",
  "version": "1.0.1",
  "maintainer": [
    "Ishaan Adarsh <ishaanad9@gmail.com>"
  ],
  "license": "postgresql",
  "provides": {
    "my_extension": {
      "file": "my_extension--1.0.1.sql",
      "version": "1.0.1"
    }
  },
  "generated_by": "Ishaan Adarsh",
  "meta-spec": {
    "version": "1.0.0",
    "url": "https://pgxn.org/meta/spec.txt"
  }
}
```
> The `meta-spec` must be version `1.0.0`.  That's referring to the version of the JSON specification, NOT your module version.  `1.0.0` is the only spec that exists.

### Step 8: Build a Distribution Package for the updated version:

> - Package your extension and the `META.json` file into a distribution package. This package should be in a compressed format (zip format) and include all the necessary files for installation and usage of your extension.

Reviewer: You should show an example of this.

### Step 9: Publish Your Upgraded Extension:

> - Use the PGXN command-line tool or web interface to publish your extension. Provide the distribution package created in the previous step and upload it to PGXN. The tool will validate the package and extract the metadata.

Reviewer: Needs examples and links, including a section on how to create a PGXN account. But again, I'd urge you to make PGXN distribution a separate lesson.

Author: i will make it  separate lesson, with pictures and link to docs.
